File: app/src/main/java/com/example/myapplication/feature/progress/ProgressViewModel.kt
package com.example.myapplication.feature.progress

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.myapplication.core.model.*
import com.example.myapplication.core.program.ProgramSelectionResult
import com.example.myapplication.core.program.ProgramSelector
import com.example.myapplication.core.progress.ProgressCalculator
import com.example.myapplication.core.progress.WeeklyInsightEngine
import com.example.myapplication.core.progress.GoalForecast
import com.example.myapplication.core.progress.GoalForecastCalculator
import com.example.myapplication.data.WorkoutRepository
import com.example.myapplication.data.WorkoutFeedbackRepository
import com.example.myapplication.core.feedback.WorkoutFeedback
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth
import java.time.temporal.TemporalAdjusters
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.launch
import kotlinx.coroutines.ExperimentalCoroutinesApi

@OptIn(ExperimentalCoroutinesApi::class)
class ProgressViewModel(
    repository: WorkoutRepository,
    private val programs: List<ProgramTemplate> = emptyList(),
    private val exercises: List<ExerciseDefinition> = emptyList(),
    private val feedbackRepository: WorkoutFeedbackRepository? = null,
    private val currentEpochDay: () -> Long,
) : ViewModel() {
    private val today = MutableStateFlow(currentEpochDay())
    private val selectedMonth = MutableStateFlow(YearMonth.from(LocalDate.ofEpochDay(today.value)))
    private val _uiState = MutableStateFlow<ProgressUiState>(ProgressUiState.Loading)
    val uiState: StateFlow<ProgressUiState> = _uiState.asStateFlow()

    private val exercisesCatalog = exercises.associateBy { it.id }

    private data class ProgressInputs(
        val goal: ActiveGoal?,
        val completed: List<CompletedWorkout>,
        val history: List<WorkoutHistoryEntry>,
        val feedback: List<WorkoutFeedback>,
    )

    init {
        viewModelScope.launch {
            val feedbackFlow = repository.observeActiveGoal().flatMapLatest { goal ->
                if (goal == null || feedbackRepository == null) flowOf(emptyList())
                else feedbackRepository.observeForGoal(goal.id)
            }
            val inputs = combine(
                repository.observeActiveGoal(),
                repository.observeCompletedWorkouts(),
                repository.observeWorkoutHistory(),
                feedbackFlow,
            ) { goal, completed, history, feedback ->
                ProgressInputs(goal, completed, history, feedback)
            }
            combine(
                inputs,
                today,
                selectedMonth,
            ) { data, currentDay, month ->
                resolve(data.goal, data.completed, data.history, data.feedback, currentDay, month)
            }
                .collect { _uiState.value = it }
        }
    }

    fun refreshToday() {
        today.value = currentEpochDay()
    }

    fun previousMonth() = moveMonth(-1)
    fun nextMonth() = moveMonth(1)

    private fun moveMonth(delta: Long) {
        selectedMonth.value = runCatching { selectedMonth.value.plusMonths(delta) }
            .getOrElse { selectedMonth.value }
    }

    private fun resolve(
        goal: ActiveGoal?,
        completedHistory: List<CompletedWorkout>,
        workoutHistory: List<WorkoutHistoryEntry>,
        feedback: List<WorkoutFeedback>,
        currentDay: Long,
        month: YearMonth,
    ): ProgressUiState {
        val allDates = completedHistory.map { it.completedEpochDay }
        val marked = allDates.asSequence()
            .filter { YearMonth.from(LocalDate.ofEpochDay(it)) == month }
            .toSet()
        val monthCount = ProgressCalculator.completedSessionsByMonth(allDates)[month] ?: 0
        val canPrevious = month != YearMonth.of(-999_999_999, 1)
        val canNext = month != YearMonth.of(999_999_999, 12)

        // 1. Calculate Weekly Stats (last 4 weeks)
        val currentMonday = LocalDate.ofEpochDay(currentDay)
            .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        val weeklyStats = (3 downTo 0).map { weeksAgo ->
            val monday = currentMonday.minusWeeks(weeksAgo.toLong())
            val sunday = monday.plusDays(6)
            val count = allDates.count { it in monday.toEpochDay()..sunday.toEpochDay() }
            val label = when (weeksAgo) {
                0 -> "Tuần này"
                1 -> "Tuần trước"
                else -> "$weeksAgo tuần trước"
            }
            WeeklyCompletedStats(label, count)
        }

        if (goal == null) {
            return ProgressUiState.NoActiveGoal(
                selectedMonth = month,
                markedEpochDays = marked,
                completedInMonth = monthCount,
                canNavigatePrevious = canPrevious,
                canNavigateNext = canNext,
                weeklyStats = weeklyStats,
                muscleStats = emptyList()
            )
        }

        val activeDates = completedHistory.asSequence()
            .filter { it.goalId == goal.id }
            .map { it.completedEpochDay }
            .toList()

        // 2. Calculate Muscle Stats (based on active goal program sequence)
        val muscleStats = mutableMapOf<MuscleGroup, Int>()
        val programSelection = ProgramSelector.select(goal.config, programs)
        if (programSelection is ProgramSelectionResult.Found) {
            val program = programSelection.program
            // Workouts completed correspond to the first N workouts in sequence
            val completedCount = activeDates.size
            val completedWorkouts = program.workouts.sortedBy { it.sequence }.take(completedCount)
            for (workout in completedWorkouts) {
                for (presc in workout.exercises) {
                    val definition = exercisesCatalog[presc.exerciseId]
                    if (definition != null) {
                        muscleStats[definition.primaryMuscle] = (muscleStats[definition.primaryMuscle] ?: 0) + 1
                    }
                }
            }
        }
        val sortedMuscleStats = muscleStats.map { MuscleCompletedStats(it.key, it.value) }
            .sortedByDescending { it.count }
        val weeklyInsights = WeeklyInsightEngine.generate(
            history = workoutHistory.filter { it.goalId == goal.id },
            feedback = feedback,
            todayEpochDay = currentDay,
        )
        val activeHistory = workoutHistory.filter { it.goalId == goal.id }
        val firstDue = activeHistory.minOfOrNull { it.dueEpochDay }
        val finalDue = activeHistory.maxOfOrNull { it.dueEpochDay }
        val elapsedWeeks = firstDue?.let { due ->
            runCatching { Math.subtractExact(currentDay, due).coerceAtLeast(0L) / 7L }.getOrDefault(0L)
        } ?: 0L
        val forecast = if (firstDue == null || finalDue == null) {
            GoalForecast.InsufficientData
        } else {
            runCatching {
                GoalForecastCalculator.calculate(
                    totalSessions = goal.totalWorkouts,
                    completedSessions = activeDates.size,
                    sessionsPerWeek = goal.config.sessionsPerWeek,
                    firstDueEpochDay = firstDue,
                    plannedFinalDueEpochDay = finalDue,
                    todayEpochDay = currentDay,
                )
            }.getOrDefault(GoalForecast.InsufficientData)
        }

        return ProgressUiState.Content(
            percentage = ProgressCalculator.percentage(activeDates.size, goal.totalWorkouts),
            completedActive = activeDates.size,
            totalActive = goal.totalWorkouts,
            weeklyStreak = ProgressCalculator.weeklyStreak(activeDates, goal.config.sessionsPerWeek, currentDay),
            targetPerWeek = goal.config.sessionsPerWeek,
            selectedMonth = month,
            markedEpochDays = marked,
            completedInMonth = monthCount,
            canNavigatePrevious = canPrevious,
            canNavigateNext = canNext,
            weeklyStats = weeklyStats,
            muscleStats = sortedMuscleStats,
            weeklyInsights = weeklyInsights,
            goalForecast = forecast,
            forecastCompletedSessions = activeDates.size,
            forecastElapsedWeeks = elapsedWeeks,
        )
    }
}

File: flutter/lib/feature/progress/progress_view_model.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../core/model/workout_models.dart';
import '../../core/model/catalog_models.dart';
import '../../core/model/feedback_models.dart';
import '../../core/progress/progress_calculator.dart';
import '../../core/progress/weekly_insight_engine.dart';
import '../../core/progress/goal_forecast_calculator.dart';
import '../../core/program/program_selector.dart';
import '../../data/local/database.dart';
import '../../data/providers/data_providers.dart';
import 'progress_ui_state.dart';

final progressEpochDayProvider = Provider<int>((ref) {
  return currentLocalEpochDay();
});

class ProgressSelectedMonthNotifier extends Notifier<YearMonth> {
  @override
  YearMonth build() => YearMonth.fromEpochDay(ref.watch(progressEpochDayProvider));

  void update(YearMonth Function(YearMonth) cb) => state = cb(state);
}

final progressSelectedMonthProvider = NotifierProvider<ProgressSelectedMonthNotifier, YearMonth>(
  ProgressSelectedMonthNotifier.new,
);

class ProgressWeightFilterNotifier extends Notifier<WeightFilter> {
  @override
  WeightFilter build() => WeightFilter.last30Days;

  void set(WeightFilter val) => state = val;
}

final progressWeightFilterProvider = NotifierProvider<ProgressWeightFilterNotifier, WeightFilter>(
  ProgressWeightFilterNotifier.new,
);

// Input streams
final progressActiveGoalProvider = StreamProvider<ActiveGoal?>((ref) {
  return ref.watch(workoutRepositoryProvider).observeActiveGoal();
});

final progressCompletedWorkoutsProvider = StreamProvider<List<CompletedWorkout>>((ref) {
  return ref.watch(workoutRepositoryProvider).observeCompletedWorkouts();
});

final progressWorkoutHistoryProvider = StreamProvider<List<WorkoutHistoryEntry>>((ref) {
  return ref.watch(workoutRepositoryProvider).observeWorkoutHistory();
});

final progressWeightHistoryProvider = StreamProvider<List<WeightMeasurement>>((ref) {
  final db = ref.watch(gymDatabaseProvider);
  return db.personalizationDao.observeWeightHistory();
});

final progressFeedbackProvider = StreamProvider<List<WorkoutFeedback>>((ref) {
  final activeGoalAsync = ref.watch(progressActiveGoalProvider);
  final activeGoal = activeGoalAsync.value;
  if (activeGoal == null) return Stream.value(const []);
  final feedbackRepo = ref.watch(workoutFeedbackRepositoryProvider);
  return feedbackRepo.observeForGoal(activeGoal.id);
});

// UI State provider combining all inputs
final progressUiStateProvider = Provider<ProgressUiState>((ref) {
  final activeGoalAsync = ref.watch(progressActiveGoalProvider);
  final completedAsync = ref.watch(progressCompletedWorkoutsProvider);
  final historyAsync = ref.watch(progressWorkoutHistoryProvider);
  final weightsAsync = ref.watch(progressWeightHistoryProvider);
  final feedbackAsync = ref.watch(progressFeedbackProvider);

  final selectedMonth = ref.watch(progressSelectedMonthProvider);
  final weightFilter = ref.watch(progressWeightFilterProvider);

  final activeGoal = activeGoalAsync.value;
  final completed = completedAsync.value;
  final history = historyAsync.value;
  final weights = weightsAsync.value;
  final feedback = feedbackAsync.value;

  if (completed == null || history == null || weights == null) {
    return const ProgressUiStateLoading();
  }

  final allDates = completed.map((w) => w.completedEpochDay).toList();
  final allDatesSet = allDates.toSet();

  final marked = allDates
      .where((d) => YearMonth.fromEpochDay(d) == selectedMonth)
      .toSet();

  final monthCount = ProgressCalculator.completedSessionsByMonth(allDates)[selectedMonth] ?? 0;

  // 1. Calculate Weekly Stats (last 4 weeks)
  final todayDay = ref.watch(progressEpochDayProvider);
  final todayDate = DateTime.fromMillisecondsSinceEpoch(todayDay * 24 * 60 * 60 * 1000, isUtc: true);
  
  // Find Monday of current week
  final currentMonday = todayDate.subtract(Duration(days: todayDate.weekday - 1));
  final currentMondayEpoch = currentMonday.millisecondsSinceEpoch ~/ (24 * 60 * 60 * 1000);

  final weeklyStats = List.generate(4, (index) {
    final weeksAgo = 3 - index;
    final monday = currentMondayEpoch - (weeksAgo * 7);
    final sunday = monday + 6;
    final count = allDates.where((d) => d >= monday && d <= sunday).length;
    final label = weeksAgo == 0
        ? "Tuần này"
        : weeksAgo == 1
            ? "Tuần trước"
            : "$weeksAgo tuần trước";
    return WeeklyCompletedStats(weekLabel: label, count: count);
  });

  // Filter weights
  final filteredWeights = weights
      .where((w) => w.epochDay >= todayDay - weightFilter.days)
      .toList()
    ..sort((a, b) => a.epochDay.compareTo(b.epochDay));

  if (activeGoal == null) {
    return ProgressUiStateNoActiveGoal(
      selectedMonth: selectedMonth,
      markedEpochDays: marked,
      completedInMonth: monthCount,
      canNavigatePrevious: true,
      canNavigateNext: true,
      weeklyStats: weeklyStats,
      muscleStats: const [],
      weightHistory: filteredWeights,
      weightFilter: weightFilter,
      allCompletedDates: allDatesSet,
    );
  }

  final activeDates = completed
      .where((w) => w.goalId == activeGoal.id)
      .map((w) => w.completedEpochDay)
      .toList();

  // 2. Calculate Muscle Stats (based on active goal program sequence)
  final catalog = ref.watch(assetCatalogRepositoryProvider);
  final programs = catalog.programs;
  final exercisesMap = {for (var e in catalog.exercises) e.id: e};

  final muscleStatsMap = <MuscleGroup, int>{};
  final selection = ProgramSelector.select(activeGoal.config, programs);
  if (selection is ProgramSelectionFound) {
    final program = selection.program;
    final completedCount = activeDates.length;
    final completedSessions = (List<WorkoutTemplate>.from(program.workouts)
          ..sort((a, b) => a.sequence.compareTo(b.sequence)))
        .take(completedCount);

    for (var session in completedSessions) {
      for (var presc in session.exercises) {
        final definition = exercisesMap[presc.exerciseId];
        if (definition != null) {
          muscleStatsMap[definition.primaryMuscleGroup] =
              (muscleStatsMap[definition.primaryMuscleGroup] ?? 0) + 1;
        }
      }
    }
  }

  final muscleStats = muscleStatsMap.entries
      .map((entry) => MuscleCompletedStats(muscleGroup: entry.key, count: entry.value))
      .toList()
    ..sort((a, b) => b.count.compareTo(a.count));

  final weeklyInsights = WeeklyInsightEngine.generate(
    history: history.where((h) => h.goalId == activeGoal.id).toList(),
    feedback: feedback ?? const [],
    todayEpochDay: todayDay,
  );

  final activeHistory = history.where((h) => h.goalId == activeGoal.id).toList();
  final firstDue = activeHistory.isEmpty
      ? null
      : activeHistory.map((h) => h.dueEpochDay).reduce((a, b) => a < b ? a : b);
  final finalDue = activeHistory.isEmpty
      ? null
      : activeHistory.map((h) => h.dueEpochDay).reduce((a, b) => a > b ? a : b);

  final elapsedWeeks = firstDue == null ? 0 : ((todayDay - firstDue).clamp(0, double.infinity).toInt() ~/ 7);

  final forecast = (firstDue == null || finalDue == null)
      ? GoalForecastInsufficientData()
      : GoalForecastCalculator.calculate(
          totalSessions: activeGoal.totalWorkouts,
          completedSessions: activeDates.length,
          sessionsPerWeek: activeGoal.config.sessionsPerWeek,
          firstDueEpochDay: firstDue,
          plannedFinalDueEpochDay: finalDue,
          todayEpochDay: todayDay,
        );

  return ProgressUiStateContent(
    percentage: ProgressCalculator.percentage(activeDates.length, activeGoal.totalWorkouts),
    completedActive: activeDates.length,
    totalActive: activeGoal.totalWorkouts,
    weeklyStreak: ProgressCalculator.weeklyStreak(
      completedEpochDays: activeDates,
      targetPerWeek: activeGoal.config.sessionsPerWeek,
      currentEpochDay: todayDay,
    ),
    targetPerWeek: activeGoal.config.sessionsPerWeek,
    selectedMonth: selectedMonth,
    markedEpochDays: marked,
    completedInMonth: monthCount,
    canNavigatePrevious: true,
    canNavigateNext: true,
    weeklyStats: weeklyStats,
    muscleStats: muscleStats,
    weeklyInsights: weeklyInsights,
    goalForecast: forecast,
    forecastCompletedSessions: activeDates.length,
    forecastElapsedWeeks: elapsedWeeks,
    weightHistory: filteredWeights,
    weightFilter: weightFilter,
    allCompletedDates: allDatesSet,
  );
});
